fdbserver/DataDistribution.actor.cpp

```diff
@@ -2569,6 +2585,188 @@ ACTOR Future<Void> waitUntilHealthy(DDTeamCollection* self, double extraDelay =
 	}
 }
 
+// Take a snapshot of necessary data structures from `DDTeamCollection` and print them out with yields to avoid slow
+// task on the run loop. This function is very expensive to call due to the data it needs to copy. Use with caution.
+ACTOR Future<Void> printSnapshotTeamsInfo(Reference<DDTeamCollection> self, std::string tcIdString) {
+	state DatabaseConfiguration configuration;
+	state std::map<UID, Reference<TCServerInfo>> server_info;
+	state std::map<UID, ServerStatus> server_status;
+	state vector<Reference<TCTeamInfo>> teams;
+	state std::map<Standalone<StringRef>, Reference<TCMachineInfo>> machine_info;
+	state std::vector<Reference<TCMachineTeamInfo>> machineTeams;
+	// state std::vector<std::string> internedLocalityRecordKeyNameStrings;
+	// state int machineLocalityMapEntryArraySize;
+	// state std::vector<Reference<LocalityRecord>> machineLocalityMapRecordArray;
+	state int traceEventsPrinted = 0;
+	state std::vector<const UID*> serverIDs;
+	loop {
+		wait(self->printDetailedTeamsInfo.onTrigger() || delay(SERVER_KNOBS->DD_TEAMS_INFO_PRINT_INTERVAL));
+
+		traceEventsPrinted = 0;
+		if (now() - self->lastTeamCollectionInfoSnapshotTime >= SERVER_KNOBS->DD_TEAMS_INFO_SNAPSHOT_REFRESH_INTERVAL) {
```

Comment:
I don't think we need to distinguish the `DD_TEAMS_INFO_PRINT_INTERVAL ` and `DD_TEAMS_INFO_SNAPSHOT_REFRESH_INTERVAL `. 
We can remove DD_TEAMS_INFO_SNAPSHOT_REFRESH_INTERVAL.

Whenever we want to print, we should take a new snapshot; because when print is triggered by an event, we should always print the latest info. 
We don't take a snapshot when we don't need to print.

Reply:
> Whenever we want to print, we should take a new snapshot; because when print is triggered by an event, we should always print the latest info.

I partially agree, but this would require the caller who sends the trigger to be aware of the cost and limit the frequency of the trigger, which in my experience is not 100% reliable.

Reply:
about "limit the frequency of the trigger", can we limit the frequency of the snapshot and printing in this actor?
if (now() - self->lastTimeTCInfoDump) < SERVER_KNOBS->DD_TC_INFO_DUMP_DELAY) {
     continue; // and wait for the next trigger.
}

Reply:
> I think the real overhead is the trace printing.

After the change that disabled the copy of locality related data structures, which contains lots strings, I agree.